set/name_trie.go

```diff
@@ -0,0 +1,96 @@
+package set
+
+import (
+    "sort"
+)
+
+type NameTrie struct {
+    children   map[rune]*NameTrie
+    terminates bool
+}
+
+func createNameTrie() *NameTrie {
+    return &NameTrie{make(map[rune]*NameTrie), false}
+}
+
+func (tree *NameTrie) Exists(name string) bool {
+    nameSlice := []rune(name)
+    node, ok := tree.traverse(nameSlice, false)
+    return ok && node.terminates
+}
+
+func (tree *NameTrie) Insert(name string) {
+    nameSlice := []rune(name)
+    node, ok := tree.traverse(nameSlice, true)
+    if ok {
+        node.terminates = true
+    }
+}
+
+func (tree *NameTrie) Delete(name string) {
+    nameSlice := []rune(name)
+    node, ok := tree.traverse(nameSlice, false)
+    if ok {
+        node.terminates = false
+    }
+}
+
+func (node *NameTrie) OrderedChildren() []rune {
+    keys := []string{}
+    for suffix := range(node.children){ keys = append(keys, string(suffix)) }
+    sort.Strings(keys)
+    sorted_runes := []rune{}
+    for i := range(keys){ sorted_runes = append(sorted_runes, rune(keys[i][0]))}
+    return sorted_runes
+}
+
+func (tree *NameTrie) traverse(remainder []rune, create bool) (nextTree *NameTrie, ok bool) {
+    nextRune := remainder[0]
+    nextTree, ok = tree.children[nextRune]
+    if !ok {
+        if create {
+            tree.children[nextRune] = createNameTrie()
+            nextTree = tree.children[nextRune]
+        } else {
+            return createNameTrie(), false
+        }
+    }
+    if len(remainder) < 2 {
+        return nextTree, true
+    } else {
+        return nextTree.traverse(remainder[1:], create)
+    }
+}
+
+func (tree *NameTrie) FirstName(name string) (string, bool){
+    nameSlice := []rune(name)
+    node, ok := tree.traverse(nameSlice, false)
+    searchingName, nameFound := "", false
+    if ok {
+        c := make(chan string)
+        go func(){
```

Comment:
Helper libraries should avoid using goroutines unless the problem is inherently distributed. At minimum, should provide a procedural (not sure if that's the right word?) implementation first, then an optional concurrent implementation on top of it.

Generally, goroutines should be used at the highest level possible, ideally by the top-level consumer. Surprise goroutines buried deep in libraries tend to complicate things.

Also, for the sake of making your own life easier. Go makes concurrency easy, but it's still much easier for humans (and tests) to reason about non-concurrent code, make sure we got it right, then add concurrency to it later.

And finally, performance-wise, I suspect goroutines here actually hurt more than help for the shape of data that we'll be dealing with (5-20 chars, 10-100 entries).